ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -942,6 +899,323 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+  Acquiring a ledger view for RollForward validation
+-------------------------------------------------------------------------------}
+
+data LedgerViewCheck blk =
+    -- | The upstream chain no longer intersects with our current chain because
+    -- our current chain changed in the background.
+    NoLongerIntersects
+    -- | The upstream chain still intersects with our chain, return the latest
+    -- 'KnownIntersectionState' and the 'LedgerView' necessary to validate the
+    -- RollForward header.
+  | LedgerViewAcquired
+      (KnownIntersectionState blk)
+      (Ticked (LedgerView (BlockProtocol blk)))
+
+-- | The peer has sent a RollForward header but their chain is not dense enough
+-- for us to validate the header in the usual way (via forecasting). We'll have
+-- to pre-emptively fetch some of their blocks in order to do so.
+--
+-- This is the state type for the logic requests fetchs and processes the
+-- fetched blocks.
+data LowDensity blk
+  = LowDensity
+      !(KnownIntersectionState blk)
+      -- ^ The 'mostRecentIntersection' via which we obtain the
+      -- initially-unadvanced 'LedgerState' (whose 'LedgerView' forecast did
+      -- not reach the slot of the RollForward header).
+      !(LedgerState blk)
+      -- ^ The advanced ledger state so far.
+      ![RealPoint blk]
+      -- ^ The remaining blocks on the peer's chain to fetch and then apply in
+      -- order to further advance this ledger state.
+      !RequestStatus
+      -- ^ Whether the candidate fragment is certainly set to the next point.
+  deriving (Generic)
+  deriving anyclass (NoThunks)
+
+data RequestStatus = RequestReady | RequestStale
+  deriving (Generic, Eq)
+  deriving anyclass (NoThunks)
+
+-- | Attempt to forecast the necessary ledger view from the
+-- 'mostRecentIntersection'. If that doesn't work, enter the supplied
+-- 'LowDensity' handler.
+checkDensityOrContinue :: forall blk m.
+     (MonadSTM m, LedgerSupportsProtocol blk)
+  => TopLevelConfig blk
+  -> ChainDbView m blk
+  -> SlotNo
+     -- ^ forecast to this slot
+  -> Stateful (STM m) (LowDensity blk) (m (LedgerViewCheck blk))
+     -- ^ how to handle a low density chain
+  -> Stateful (STM m) (KnownIntersectionState blk) (m (LedgerViewCheck blk))
+checkDensityOrContinue cfg cdbView hdrSlot kont = Stateful $ \kis -> do
+    let KnownIntersectionState {
+            mostRecentIntersection
+          , theirFrag
+          } = kis
+
+        theirSuffix :: AnchoredFragment (Header blk)
+        theirSuffix =
+            case AF.splitAfterPoint theirFrag mostRecentIntersection of
+              Nothing              -> error "impossible!"
+              Just (_upto, suffix) -> suffix
+
+    st <- getIntersectionLedgerState cdbView kis
+    case tryForecastLedgerView (Proxy @blk) lCfg st hdrSlot of
+      Just v  -> return $ return $ LedgerViewAcquired kis v
+      Nothing -> do
+          -- If their suffix is so long that the normal ChainSync/BlockFetch
```

Comment:
Is this just an optimization? Doesn't seem worthwhile. Them having more than k blocks after the intersection is an extremely rare case.